File: zero_task_java/src/include/Parser.java
package include;

public class Parser {

    private String expression;
    private int index = 0;

    private void make_result(Result current_expression) {
        switch (current_expression.number) {
            case 2: {
                System.out.print("(" + current_expression.expression + ",");
                make_result(current_expression.left);
                System.out.print(",");
                make_result(current_expression.right);
                System.out.print(")");
                break;
            }
            case 1: {
                System.out.print("(!");
                make_result(current_expression.left);
                System.out.print(")");
                break;
            }
            case 0: {
                System.out.print(current_expression.expression);
                break;
            }
            default:
                break;
        }
    }

    private Result parsing() {
        return binary_parsing(0);
    }

    private Result binary_parsing(int current_priority) {
        Result left = expressions_parsing();

        while (true) {
            String op = char_parsing();
            int priority = get_priority(op);
            if (priority == 1) {
                if (priority < current_priority) {
                    index -= op.length();
                    return left;
                }
            } else {
                if (priority <= current_priority) {
                    index -= op.length();
                    return left;
                }
            }

            Result right = binary_parsing(priority);
            left = new Result(op, left, right);
        }
    }

    private Result expressions_parsing() {
        String currentChar = char_parsing();

        if (currentChar.equals("(")) {
            Result  result = parsing();
            index++;
            return result;
        }

        if (currentChar.charAt(0) >= 65 && currentChar.charAt(0) <= 90) {
            return new Result(currentChar);
        }

        return new Result(currentChar, expressions_parsing());
    }

    private String char_parsing() {

        if (index == expression.length()) {
            return "";
        }

        if (expression.charAt(index) >= 65 && expression.charAt(index) <= 90) {
            StringBuilder number = new StringBuilder();
            while (index < expression.length() && ((expression.charAt(index) >= 65 && expression.charAt(index) <= 90) ||
                    (expression.charAt(index) >= 48 && expression.charAt(index) <= 57))) {
                number.append(expression.charAt(index++));
            }
            return number.toString();
        }

        String[] operations = {"(", ")", "->", "&", "|", "!"};
        for (String operation :operations){
            if (operation.charAt(0) == expression.charAt(index)) {
                index += operation.length();
                return operation;
            }
        }
        return "";
    }

    private int get_priority(String operation) {
        if (operation.equals("->")) {
            return 1;
        }
        if (operation.equals("|")) {
            return 2;
        }
        if (operation.equals("&")) {
            return 3;
        }
        if (operation.equals("!")) {
            return 4;
        }
        return 0;
    }

    public Parser(String expression) {
        this.expression = expression;
    }

    ;

    public void parse() {
        index = 0;
        make_result(parsing());
    }
}

File: zeroTaskKotlin/src/include/Parser.kt
package include

class Parser(private val expression: String) {
    private var index = 0

    private fun makeResult(current_expression: Result) {
        when (current_expression.number) {
            2 -> {
                print("(" + current_expression.expression + ",")
                makeResult(current_expression.left!!)
                print(",")
                makeResult(current_expression.right!!)
                print(")")
            }
            1 -> {
                print("(!")
                makeResult(current_expression.left!!)
                print(")")
            }
            0 -> {
                print(current_expression.expression)
            }
            else -> {
            }
        }
    }

    private fun parsing(): Result {
        return binaryParsing(0)
    }

    private fun binaryParsing(current_priority: Int): Result {
        var left = expressionsParsing()

        while (true) {
            val op = charParsing()
            val priority = getPriority(op)
            if (priority == 1) {
                if (priority < current_priority) {
                    index -= op.length
                    return left
                }
            } else {
                if (priority <= current_priority) {
                    index -= op.length
                    return left
                }
            }

            val right = binaryParsing(priority)
            left = Result(op, left, right)
        }
    }

    private fun expressionsParsing(): Result {
        val currentChar = charParsing()

        if (currentChar == "(") {
            val result = parsing()
            index++
            return result
        }

        return if (currentChar[0].toInt() in 65..90) {
            Result(currentChar)
        } else Result(currentChar, expressionsParsing())

    }

    private fun charParsing(): String {

        if (index == expression.length) {
            return ""
        }

        if (expression[index].toInt() in 65..90) {
            val number = StringBuilder()
            while (index < expression.length && (expression[index].toInt() in 65..90 || expression[index].toInt() in 48..57)) {
                number.append(expression[index++])
            }
            return number.toString()
        }

        val operations = arrayOf("(", ")", "->", "&", "|", "!")
        for (operation in operations) {
            if (operation[0] == expression[index]) {
                index += operation.length
                return operation
            }
        }
        return ""
    }

    private fun getPriority(operation: String): Int {
        if (operation == "->") {
            return 1
        }
        if (operation == "|") {
            return 2
        }
        if (operation == "&") {
            return 3
        }
        return if (operation == "!") {
            4
        } else 0
    }

    fun parse() {
        index = 0
        makeResult(parsing())
    }
}
